### PR TITLE
Block Editor: Add BlockContext component to type-checking

### LINKED ISSUE #21666: Block Editor: BlockContext: Opt-in BlockContext component to type checking
Previously: #21467 (https://github.com/WordPress/gutenberg/pull/21467#issuecomment-614832779)
Related: #18838
Blocked by: #21613

**Context:**

>I'd hoped to include a bit of extra type-checking for the new `@wordpress/block-editor` file in f95972d. Unfortunately, while this worked in my (stale) local branch, dependent types were removed in #21613 as a result of some issues with this week's packages publishing.

**Task:**

Once the issue associated with #21613 is resolved, reintroduce the changes of 0ff9acda2dee96bf6f571aec27cb917ef664db05.

Note: This commit should have also included the following additional revision:

```diff
diff --git a/packages/block-editor/tsconfig.json b/packages/block-editor/tsconfig.json
index 7cd3fd793..61d1d214a 100644
--- a/packages/block-editor/tsconfig.json
+++ b/packages/block-editor/tsconfig.json
@@ -4,6 +4,9 @@
 		"rootDir": "src",
 		"declarationDir": "build-types"
 	},
+	"references": [
+		{ "path": "bin" },
+	],
 	// NOTE: This package is being progressively typed. You are encouraged to
 	// expand this array with files which can be type-checked. At some point in
 	// the future, this can be simplified to an `includes` of `src/**/*`.
```

It is necessary for all packages referenced by a file to be included in `references`.

Related documentation: https://github.com/WordPress/gutenberg/blob/master/packages/README.md#using-typescript

### PR DESCRIPTION
**Pull request by aduth**
Closes #21666

This pull request seeks to opt in the BlockContext component (introduced in #21467) to be type-checked. It was previously planned for #21467 but removed due to blocking incompatibilities of #21767. These incompatibilities have since been resolved in #21781, so the type-checking can once again be enabled.

**Testing Instructions:**

Ensure type checking passes:

```
npm run build:package-types
```

### CODE REVIEW

**Review by youknowriad — APPROVED** [review]

**Comment by aduth** [comment]
The build caught an error, which I wasn't seeing locally when running `npm run build:package-types`.

Eventually I could reproduce it by first forcibly removing existing `build-types` folders and `.tsbuildinfo` files:

```
rm -r packages/**/build-types packages/**/*.tsbuildinfo
```

It seems there may have been some caching that was preventing the error from being logged.

It's clear enough what the solution is ([including the dependencies in `references`](https://github.com/WordPress/gutenberg/blob/master/packages/README.md#typescript)), but definitely lose a bit of trust in `npm run build:package-types` to be reliable. At least it's caught in continuous integration.

cc @sirreal

**Comment by sirreal** [comment]
> The build caught an error, which I wasn't seeing locally when running `npm run build:package-types`.

The error was that the reference omitted? This results in the missing declarations when building a dependent module?

> Eventually I could reproduce it by first forcibly removing existing `build-types` folders and `.tsbuildinfo` files:
> 
> ```
> rm -r packages/**/build-types packages/**/*.tsbuildinfo
> ```

Aside: This is effective. It should be roughly analogous to `npm run clean:package-types` (which runs `tsc --build --clean`).

> It seems there may have been some caching that was preventing the error from being logged.
> 
> It's clear enough what the solution is ([including the dependencies in `references`](https://github.com/WordPress/gutenberg/blob/master/packages/README.md#typescript)), but definitely lose a bit of trust in `npm run build:package-types` to be reliable. At least it's caught in continuous integration.

The references are the only way `tsc` knows how projects interdepend. It sounds like you had a types built locally which `tsc` was able to find and use. There's nothing like auto-discovery via `lerna` or `package.json` for TypeScript references as far as I know.

This has popped up a few times. We could probably lint for this by verifying that `references` aligns with `@wordpress/*` file dependencies in packages' `package.json`. I'm undecided how much value that would add.

File: packages/block-editor/tsconfig.json [hunk 1]
```diff
@@ -4,10 +4,14 @@
 		"rootDir": "src",
 		"declarationDir": "build-types"
 	},
+	"references": [
+		{ "path": "../element" }
+	],
```
Comment (sirreal) [inline packages/block-editor/tsconfig.json:9]: This actually depends on a bunch of things:
https://github.com/WordPress/gutenberg/blob/e00a4850737c89860d98f5830cf2b0b09bc0d210/packages/block-editor/package.json#L26-L48

We should add the packages that are typed already at least.

Comment (aduth) [reply packages/block-editor/tsconfig.json:9]: > This actually depends on a bunch of things:

That package as a whole does, sure. Currently we only type-check two files, for which this `references` set is accurate. I guess it can save us some future work if we just go ahead and add all of the dependencies now?

**Comment by aduth** [comment]
@sirreal There may be a misunderstanding. My concern isn't that there is an error. My concern was that there was _no_ error for me locally, but then the error appeared in continuous integration. And only after the series of steps above was I finally able to reproduce the error locally.

tl;dr: I got a false negative when testing locally.

File: packages/block-editor/tsconfig.json (hunk 1 above)
Comment (sirreal) [reply packages/block-editor/tsconfig.json:9]: This is interesting to consider…

I've been thinking of the TypeScript project as corresponding to the JavaScript project, but in this case it really doesn't. It's a small subset of the block-editor package. As such, it has a subset of the dependencies.

I don't feel strongly about this one way or the other.

**Comment by sirreal** [comment]
I don't think there's a misunderstanding, but we may have different perspectives and not share the same concerns 🙂 

Here's how I'm looking at this:

- A project was misconfigured, missing a reference. This project is a bit interesting because it doesn't (and cannot) list all of it's package's dependencies: https://github.com/WordPress/gutenberg/pull/22353#discussion_r425224923 (an observation, not pointing fingers)
- TypeScript doesn't build (if missing) or rebuild (if stale) the dependencies when building.
- Locally, TypeScript checks the types by using the declaration files it finds.
- Pre-existing (and potentially stale) declarations were used for typechecking.
- CI encounters the issue because the dependency was not built before the dependent. This is because CI is not caching the declaration files at all.

It seems that this all stems from a misconfiguration that we can attempt to prevent in the future or to accept as a risk not worth mitigating because the difficulty or maintenance of the mitigation outweighs the severity.

Even on CI, it's possible that through other dependencies or simple race conditions, the dependencies would happen to be built by the time the declarations were required.

Note that there's a tradeoff here. We could clean and rebuild types every run, but I've anecdotally observed that rebuilding from scratch can take upwards of _forty seconds_ whereas reusing valid incremental builds is virtually instantaneous.

If we decide to mitigate, how could we prevent issues like this?

- Somehow enforce that every package-level `tsconfig.json` declare all the monorepo file dependencies as references. With this package in particular, we'd need some logic like "if it depends on another package that has a tsconfig, it should be listed in the references" because not all the dependencies of this package are TypeScript projects right now. Requires implementing and maintaining a script to handle this.
- Clean the TypeScript build more frequently. Here we move the tradeoff from speed to accuracy.
- We could clean and build each project in isolation:
  ```
  npm run clean:package-types
  npm run build:package-types -- packages/block-editor/
  ```
  I'd expect this to be very good at detecting this issue, but quite slow. Something like this could run as a scheduled job on CI.

I don't foresee this type of problem being very severe, but I may be wrong. This is mostly intuition. I'd expect problems of this sort either surface rather quickly or be harmless for the most part.

**Comment by aduth** [comment]
🤷 I just want a type-checking command that I can trust.
